cvmfs/file_watcher_kqueue.cc

```diff
@@ -0,0 +1,164 @@
+          } else if (current_ev.fflags & NOTE_LINK) {
+            // Hardlinked
+            event = file_watcher::kHardlinked;
+          } else if (current_ev.fflags & NOTE_WRITE) {
+            // Modified
+            event = file_watcher::kModified;
+          } else if (current_ev.fflags & NOTE_RENAME) {
+            // Renamed
+            event = file_watcher::kRenamed;
+          } else if (current_ev.fflags & NOTE_ATTRIB) {
+            // Attributes
+            event = file_watcher::kAttributes;
+          }
+          bool clear_handler = true;
+          if (event != file_watcher::kInvalid) {
+            current_record.handler_->Handle(current_record.file_path_, event,
+                                            &clear_handler);
+          } else {
+            LogCvmfs(kLogCvmfs, kLogDebug,
+                     "FileWatcherKqueue - Unknown event 0x%x\n",
+                     current_ev.fflags);
+          }
+
+          // Perform post-handling actions (i.e. remove, reset filter)
+          if (event == file_watcher::kDeleted) {
+            const std::string file_path = watch_records_[current_fd].file_path_;
+            EventHandler* handler = watch_records_[current_fd].handler_;
+            RemoveFilter(current_fd);
+            if (!clear_handler) {
+              RegisterFilter(file_path, handler);
+            }
+          }
+        } else {
+          LogCvmfs(kLogCvmfs, kLogDebug,
+                   "FileWatcherKqueue - Unknown kevent ident: %ld",
+                   current_ev.ident);
+        }
+      }
+    }
+  }
+  return true;
+}
+
+void FileWatcherKqueue::RemoveFilter(int fd) {
+  struct kevent remove_event;
+  EV_SET(&remove_event, fd, EVFILT_VNODE, EV_DELETE, NULL, 0, 0);
+  if (kevent(kq_, &remove_event, 1, NULL, 0, NULL) == -1) {
+    LogCvmfs(kLogCvmfs, kLogDebug,
+             "FileWatcherKqueue - Could not remove event filter from kqueue.");
+  }
+  close(fd);
+  watch_records_.erase(fd);
+}
+
+void FileWatcherKqueue::RegisterFilter(const std::string& file_path,
+                                       EventHandler* handler) {
+  bool done = false;
+  BackoffThrottle throttle(1000, 10000, 50000);
+  while (!done) {
+    int fd = open(file_path.c_str(), O_RDONLY);
+    if (fd == -1) {
+      LogCvmfs(kLogCvmfs, kLogDebug,
+               "FileWatcherKqueue - Cannot open file %s. Retrying.",
+               file_path.c_str());
+      throttle.Throttle();
+      continue;
+    }
+    struct kevent watch_event;
+    EV_SET(&watch_event, fd, EVFILT_VNODE, EV_ADD | EV_ENABLE | EV_CLEAR,
+           NOTE_DELETE | NOTE_WRITE | NOTE_EXTEND | NOTE_ATTRIB | NOTE_LINK |
+               NOTE_RENAME | NOTE_REVOKE,
+           0, 0);
+
+    if (kevent(kq_, &watch_event, 1, NULL, 0, NULL) == -1) {
+      LogCvmfs(kLogCvmfs, kLogDebug,
+               "FileWatcherKqueue - Could not register event with kqueue.");
+      return;
+    }
+
+    watch_records_[fd] = WatchRecord(file_path, handler);
```

Comment:
I'm not sure, are the established filters and there file descriptors cleaned up on destruction?